src/System.Security.SecureString/src/System/Security/SecureString.Windows.cs

```diff
@@ -0,0 +1,305 @@
+        private uint EncryptedBufferLength
+        {
+            [System.Security.SecurityCritical]  // auto-generated
+            get
+            {
+                Debug.Assert(_encryptedBuffer != null, "Buffer is not initialized!");
+                return _encryptedBuffer.Length;
+            }
+        }
+
+        [System.Security.SecurityCritical]  // auto-generated
+        private void AllocateBuffer(uint size)
+        {
+            _encryptedBuffer = SafeBSTRHandle.Allocate(null, size);
+            if (_encryptedBuffer.IsInvalid)
+            {
+                throw new OutOfMemoryException();
+            }
+        }
+
+        [System.Security.SecurityCritical]  // auto-generated
+        private void EnsureCapacity(ref SafeBSTRHandle decryptedBuffer, int capacity)
+        {
+            if (capacity > MaxLength)
+            {
+                throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
+            }
+
+            if (capacity <= _decryptedLength)
+            {
+                return;
+            }
+
+            SafeBSTRHandle newBuffer = SafeBSTRHandle.Allocate(null, (uint)capacity);
+
+            if (newBuffer.IsInvalid)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            SafeBSTRHandle.Copy(decryptedBuffer, newBuffer);
+            decryptedBuffer.Dispose();
+            decryptedBuffer = newBuffer;
+        }
+
+        [System.Security.SecurityCritical]  // auto-generated
+        private void ProtectMemory(SafeBSTRHandle decryptedBuffer)
+        {
+            Debug.Assert(!decryptedBuffer.IsInvalid, "Invalid buffer!");
+
+            if (_decryptedLength == 0)
+            {
+                return;
+            }
+
+            try
+            {
+                SafeBSTRHandle newEncryptedBuffer = null;
+                if (Interop.Crypt32.CryptProtectData(decryptedBuffer, out newEncryptedBuffer))
+                {
+                    _encryptedBuffer.Dispose();
+                    _encryptedBuffer = newEncryptedBuffer;
+                }
+                else
+                {
+                    throw new CryptographicException(Marshal.GetLastWin32Error());
+                }
+            }
+            finally
+            {
+                decryptedBuffer.ClearBuffer();
+            }
+        }
+
+        [System.Security.SecurityCritical]  // auto-generated
+        private SafeBSTRHandle UnProtectMemory()
+        {
+            Debug.Assert(!_encryptedBuffer.IsInvalid, "Invalid buffer!");
+
+            SafeBSTRHandle decryptedBuffer = null;
+            if (_decryptedLength == 0)
+            {
+                return _encryptedBuffer;
+            }
+
+            if (!Interop.Crypt32.CryptUnProtectData(_encryptedBuffer, out decryptedBuffer))
+            {
+                throw new CryptographicException(Marshal.GetLastWin32Error());
+            }
+
+            return decryptedBuffer;
+        }
+
+    }
```

Comment:
// -----------------------------
        // ---- PAL layer ends here ----
        // -----------------------------
? (Also for the Unix one)